Make `--doctool` locale aware

timothyqiu opened the pull request:
This makes i18n of the online class reference easier. One can then generate RST based on the localized XML (although the generator script needs further tweaks in order to work with non-latin content).

* Adds `indent(str)` to `String`:
    * Indents the (multiline) string with the given indentation.
    * This method is added in order to keep the translated XML correctly indented.
* Moves the loading of tool/doc translation into `editor/editor_translation.{h,cpp}`.
    * This will be used from both `EditorSettings` and the doc tool from `main`.
* Makes use of doc translation when generating XML class references, and setup the translation locale based on `-l LOCALE` CLI parameter.

The XML class reference won't be translated if `-l LOCALE` is not given, or when it's `-l en`. So it's backward compatible I think.

I'll backport this PR to `3.x` if it is OK.

timothyqiu commented:
Updated the PR, and now it only translates the XMLs when `-l LOCALE` is given.

It seems that the clearing of static `locale` variable when exiting `setup2()` does not do anything expect preventing others from knowing if any thing is given as `-l` parameter. I removed it so that `--doctool` can only translate when the parameter is given.

akien-mga commented on doc/classes/PhysicsServer3D.xml:
```diff
@@ -336,7 +336,7 @@
 			<description>
 				Returns the physics layer or layers a body can collide with.
--
+				-
```
This should be removed.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -35,6 +35,11 @@
+#include "core/io/compression.h"
+#include "core/io/file_access_memory.h"
+#include "core/io/translation_loader_po.h"
+#include "editor/doc_translations.gen.h"
+#include "editor/editor_translations.gen.h"
```
`core` shouldn't depend on `editor` in theory, see #53295.

So this should probably go back to `editor`, maybe in a new `editor/editor_translation.h` file?

akien-mga commented on core/string/translation.h:
```diff
@@ -157,6 +157,12 @@ class TranslationServer : public Object {
 
+#ifdef TOOLS_ENABLED
+	Vector<String> get_editor_locales() const;
+	void load_editor_tool_translations(const String &p_locale);
+	void load_editor_doc_translations(const String &p_locale);
```
```suggestion
	void load_editor_translations(const String &p_locale);
	void load_doc_translations(const String &p_locale);
```

To match the header names (especially if moved to an `editor/` file then there's no risk of confusion with project documentation).

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -3507,6 +3507,48 @@ char32_t String::unicode_at(int p_idx) const {
+	return indent;
+}
+
+String String::indent(const String &p_indent) const {
```
For the record we seem to have
```
core/io/json.h
72:     static String _make_indent(const String &p_indent, int p_size);
```
which could maybe be modified to use this instead if we're adding it to `String`.

timothyqiu replied:
This one is just making the correct level of indent string instead of indenting. Can be probably replaced by `p_indent.repeat(p_size)` :)

akien-mga commented:
Looks good overall and seems to work well :+1: 

I'd split the XML fixes to a dedicated PR so that they can be merged now while we keep iterating on the core and translation API changes.

timothyqiu committed:
Make `--doctool` locale aware

* Adds `indent(str)` to `String`:
    * Indent the (multiline) string with the given indentation.
    * This method is added in order to keep the translated XML correctly
      indented.
* Moves the loading of tool/doc translation into
  `editor/editor_translation.{h,cpp}`.
    * This will be used from both `EditorSettings` and the doc tool from
      `main`.
* Makes use of doc translation when generating XML class references, and
  setup the translation locale based on `-l LOCALE` CLI parameter.

The XML class reference won't be translated if `-l LOCALE` parameter is
not given, or when it's `-l en`.

akien-mga reviewed (approved):
Looks good to me for the editor/translation/main changes.

For the added String API, could use some additional review to make sure that it's a useful API to have here (and which should then likely be exposed to scripting). Otherwise it could be kept local to where it's needed since it's only needed by `DocTools` so far.

timothyqiu commented:
Rebased so CI won't complain about changed XMLs.

Also made `get_indent()` local to `DocTool` since it only returns the first valid indentation, which seems too limited. `indent(str)` is kept in `String` and exposed to scripting.

akien-mga commented on doc/classes/String.xml:
```diff
@@ -243,6 +243,14 @@
 		</method>
+		<method name="indent" qualifiers="const">
+			<return type="String" />
+			<argument index="0" name="indent" type="String" />
```
Maybe name it `prefix` to avoid repeating "indent"?

akien-mga commented on doc/classes/String.xml:
```diff
@@ -243,6 +243,14 @@
+			<return type="String" />
+			<argument index="0" name="indent" type="String" />
+			<description>
+				Returns a copy of the string with lines indented with [code]indent[/code].
```
```suggestion
				Returns a copy of the string with lines indented with [code]indent[/code]. For example, the string can be indented with two tabs using [code]"\t\t"[/code], or commented out with [code]"# "[/code].
```

akien-mga replied:
Can also add a reference to `indent` in the description of `dedent`.

akien-mga commented on editor/doc_tools.cpp:
```diff
@@ -37,12 +37,42 @@
+	return indent;
+}
+
+static String _translate_doc_string(const String &p_text) {
+	const String indent = _get_indent(p_text);
+	const String message = p_text.dedent().strip_edges();
+	const String translated = TranslationServer::get_singleton()->doc_translate(message, "");
+	// No need to restore stripped edges because they'll be stripped again later.
+	return translated.indent(indent);
+}
```
In general I feel like we could be doing a better job at handling the XML tab indentation of these strings. The only part which we need to preserve is the manual 4-space indentation in code blocks, but everything else should be something that we could simply ignore, and add manually in `_write_string` (as already done at least one the first line).

Currently we dedent, strip edges, indent again, strip edges again (so it removes indentation on the first line again?), and then escape. A lot of steps which could be done better IMO.

But that's not blocking for this PR.

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -3507,6 +3507,27 @@ char32_t String::unicode_at(int p_idx) const {
 	return operator[](p_idx);
 }
 
+String String::indent(const String &p_indent) const {
```
```suggestion
String String::indent(const String &p_prefix) const {
```
to match the exposed name.

akien-mga commented:
> I'll backport this PR to `3.x` if it is OK.

This should be good to backport now :+1:

akien-mga commented:
Thanks!
